src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,479 @@
+	int sum     = 0;
+	for (unsigned int i = 0; i < seek_samples_.size(); ++i) {
+		if (seek_samples_[i].empty()) {
+			continue;
+		}
+
+		sum += seek_sample_weights_[i];
+		if (r < sum) {
+			return i;
+		}
+	}
+	return 0;
+}
+DiskNoiseDevice::DiskNoiseDevice(const bool enable_disk_noise,
+                                 const std::string& spin_up_sample_path,
+                                 const std::string& spin_sample_path,
+                                 const std::vector<std::string>& seek_sample_paths,
+                                 const float& spin_volume,
+                                 const float& seek_volume, bool loop_spin_sample)
+        : spin_volume_(spin_volume),
+          seek_volume_(seek_volume),
+          spin_up_sample_(),
+          spin_sample_(),
+          current_seek_sample_(),
+          seek_samples_(),
+          seek_sample_weights_()
+{
+	enable_disk_noise_ = enable_disk_noise;
+	loop_spin_sample_  = loop_spin_sample;
+
+	if (!enable_disk_noise_) {
+		LOG_INFO("DISKNOISE: Disk noise emulation disabled");
+		return;
+	}
+
+	LoadSample(spin_up_sample_path, spin_up_sample_);
+	LoadSample(spin_sample_path, spin_sample_);
+	LoadSeekSamples(seek_sample_paths);
+
+	std::lock_guard<std::mutex> lock(device_mutex_);
+
+	// Lazily create the static mixer channel if it doesn't exist yet
+	if (!mix_channel_) {
+		MIXER_LockMixerThread();
+		mix_channel_ = MIXER_AddChannel(AudioCallback,
+		                                SampleRate,
+		                                ChannelName::DiskNoise,
+		                                {ChannelFeature::Stereo});
+		mix_channel_->Enable(true);
+		MIXER_UnlockMixerThread();
+		float vol_gain = percentage_to_gain(100);
+		mix_channel_->SetAppVolume({vol_gain, vol_gain});
+	}
+
+	active_devices_.push_back(this);
+}
+
+void DiskNoiseDevice::ActivateSpin()
+{
+	if (!enable_disk_noise_) {
+		return;
+	}
+
+	if (!mix_channel_->is_enabled) {
+		mix_channel_->Enable(true);
+	}
+
+	// Floppy spin samples can be re-started at any time
+	if (!loop_spin_sample_) {
+		// Check if the sample is still playing and don't interrupt if
+		// it does
+		if (spin_sample_.empty() || spin_pos_ + 1 < spin_sample_.size()) {
+			return;
+		}
+		// Restart spin sample
+		spin_pos_ = 0;
+	}
+}
+
+void DiskNoiseDevice::PlaySeek()
+{
+	if (!enable_disk_noise_) {
+		return;
+	}
+
+	const size_t index = ChooseWeightedSeekIndex();
+	if (index >= seek_samples_.size() || seek_samples_[index].empty()) {
+		return;
+	}
+
+	// Check if the sample is still playing and don't interrupt if it does
+	if (!current_seek_sample_.empty() &&
+	    seek_pos_ + 1 < current_seek_sample_.size()) {
+		return;
+	}
+
+	current_seek_sample_ = seek_samples_[index];
+	seek_pos_            = 0;
+}
+
+void DiskNoiseDevice::AudioCallback(const int frames)
+{
+	// stereo interleaved buffer
+	std::vector<float> out(frames * 2, 0.0f);
+
+	std::lock_guard<std::mutex> lock(device_mutex_);
```

Comment:
It would be much better to have a single mutex in a `DiskNoises` class that holds all these individual devices and orchestrates them. Then `DiskNoises` would have a single callback, and you'll just ask the individual `DiskNoise` device instances to render into that single buffer. So let's do that, please.